hypothesis-python/src/hypothesis/extra/array_api.py

```diff
@@ -445,15 +432,27 @@ def do_draw(self, data):
                         continue
                     else:
                         seen.add(val)
-                try:
-                    result[i] = val
-                except Exception as e:
-                    raise InvalidArgument(
-                        f"Could not add generated array element {val!r} "
-                        f"of type {type(val)} to array of dtype {result.dtype}."
-                    ) from e
-                self.check_set_value(val, result[i], self.elements_strategy)
+                result_obj[i] = val
                 assigned.add(i)
+                fill_mask[i] = False
+
+            try:
+                result = self.xp.asarray(result_obj, dtype=self.dtype)
+            except Exception as e:
+                f_expr = f"xp.asarray({result_obj}, dtype={self.dtype})"
+                raise InvalidArgument(f"Could not create array via {f_expr}") from e
+
+            for i, val in enumerate(result_obj):
+                val_0d = result[i]
+                if fill_mask[i] and self.unique:
```

Comment:
🤔 If we were on Python 3.10+, we could do `for val, val_0d, is_fill in zip(result_obj, result, fill_mask, strict=True):` - but without the `strict=` kwarg I think this is on net better.